From 396b9a9592234b8f71b37a650aeb7d199e01422c Mon Sep 17 00:00:00 2001
From: Marcelo Leitner <mleitner@redhat.com>
Date: Tue, 21 Oct 2014 16:16:00 -0400
Subject: [net] introduce netdev_alloc_pcpu_stats() for drivers

Message-id: <20141021161600.22888.83281.stgit@localhost.localdomain>
Patchwork-id: 97906
O-Subject: [RHEL7 net PATCH 009/124] net: introduce netdev_alloc_pcpu_stats() for drivers
Bugzilla: 1148041 1155088
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Backport of upstream commit: 1c213bd24ad04f4430031d20d740d7783162b099
Backport notes:
Just introduces netdev_alloc_pcpu_stats and doesn't update all its
possible callers, avoiding unecessary changes.

 commit 1c213bd24ad04f4430031d20d740d7783162b099
 Author: WANG Cong <xiyou.wangcong@gmail.com>
 Date:   Thu Feb 13 11:46:28 2014 -0800

    net: introduce netdev_alloc_pcpu_stats() for drivers

    There are many drivers calling alloc_percpu() to allocate pcpu stats
    and then initializing ->syncp. So just introduce a helper function for them.

    Cc: David S. Miller <davem@davemloft.net>
    Signed-off-by: Cong Wang <xiyou.wangcong@gmail.com>
    Signed-off-by: David S. Miller <davem@davemloft.net>

Signed-off-by: Marcelo Ricardo Leitner <mleitner@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/include/linux/netdevice.h b/include/linux/netdevice.h
index 4eab278..9301365 100644
--- a/include/linux/netdevice.h
+++ b/include/linux/netdevice.h
@@ -1711,6 +1711,20 @@ struct udp_offload {
  struct offload_callbacks callbacks;
 };
 
+#define netdev_alloc_pcpu_stats(type)    \
+({        \
+ typeof(type) *pcpu_stats = alloc_percpu(type);  \
+ if (pcpu_stats) {     \
+  int i;      \
+  for_each_possible_cpu(i) {   \
+   typeof(type) *stat;   \
+   stat = per_cpu_ptr(pcpu_stats, i); \
+   u64_stats_init(&stat->syncp);  \
+  }      \
+ }       \
+ pcpu_stats;      \
+})
+
 #include <linux/notifier.h>
 
 /* netdevice notifier chain. Please remember to update the rtnetlink
-- 
1.7.1